runners/local-artifact-service-java/src/main/java/org/apache/beam/artifact/local/LocalFileSystemArtifactStagerService.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.artifact.local;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.ArrayList;
+import java.util.Collection;
+import org.apache.beam.sdk.common.runner.v1.ArtifactApi.Artifact;
+import org.apache.beam.sdk.common.runner.v1.ArtifactApi.CommitManifestRequest;
+import org.apache.beam.sdk.common.runner.v1.ArtifactApi.CommitManifestResponse;
+import org.apache.beam.sdk.common.runner.v1.ArtifactApi.PutArtifactRequest;
+import org.apache.beam.sdk.common.runner.v1.ArtifactApi.PutArtifactResponse;
+import org.apache.beam.sdk.common.runner.v1.ArtifactStagingServiceGrpc.ArtifactStagingServiceImplBase;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** An {@code ArtifactStagingService} which stages files to a local temp directory. */
+public class LocalFileSystemArtifactStagerService extends ArtifactStagingServiceImplBase {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(LocalFileSystemArtifactStagerService.class);
+
+  public static LocalFileSystemArtifactStagerService create(File base) {
+    return new LocalFileSystemArtifactStagerService(base);
+  }
+
+  private final File stagingBase;
+
+  private LocalFileSystemArtifactStagerService(File stagingBase) {
+    this.stagingBase = stagingBase;
+  }
+
+  @Override
+  public StreamObserver<PutArtifactRequest> putArtifact(
+      final StreamObserver<PutArtifactResponse> responseObserver) {
+    return new StreamObserver<PutArtifactRequest>() {
+      private File destination;
+      private FileChannel target;
+
+      @Override
+      public void onNext(PutArtifactRequest value) {
+        switch (value.getContentCase()) {
```

Comment:
<!--new_thread; commit:fd3270d35ab3f8121ad44c2521dd0ade4223d84c; resolved:0-->
One way to reduce the complexity caused by this protocol would be to bring to the top level the state transition from "expecting name" to "expecting chunks". The initial StreamObserver would require a name, and from then on forward all other requests to a delegate. You'll eliminate the switch and gain immutability. Basically, it is a service-side model of the more naive two-round-trip version of this service.

Reply:
+1